tests/functions.rs

```diff
@@ -378,3 +378,121 @@ compile_pass! {
     ),
     ( )
 }
+
+compile_pass! {
+    num_min_max,
+    "
+       fn u64_min() -> u64;
+       fn u64_max() -> u64;
+       fn i64_min() -> i64;
+       fn i64_max() -> i64;
+       fn u32_min() -> u32;
+       fn u32_max() -> u32;
+       fn i32_min() -> i32;
+       fn i32_max() -> i32;
+       fn i32_identity(v: i32) -> i32;
+       fn i64_identity(v: i64) -> i64;
+       fn u32_identity(v: u32) -> u32;
+       fn u64_identity(v: u64) -> u64;
+    ",
+    (
+        pub fn u64_min() -> u64 {
+            u64::MIN
+        }
+        pub fn u64_max() -> u64 {
+            u64::MAX
+        }
+        pub fn i64_min() -> i64 {
+            i64::MIN
+        }
+        pub fn i64_max() -> i64 {
+            i64::MAX
+        }
+
+        pub fn u32_min() -> u32 {
+            u32::MIN
+        }
+        pub fn u32_max() -> u32 {
+            u32::MAX
+        }
+        pub fn i32_min() -> i32 {
+            i32::MIN
+        }
+        pub fn i32_max() -> i32 {
+            i32::MAX
+        }
+
+        pub fn i32_identity(v: i32) -> i32 { v }
+        pub fn i64_identity(v: i64) -> i64 { v }
+        pub fn u32_identity(v: u32) -> u32 { v }
+        pub fn u64_identity(v: u64) -> u64 { v }
+    ),
+    (),
+    (
+        var x = api.u64_min();
+        assert(x == 0);
+        assert(api.u64_identity(x) == x);
+
+        x = api.u64_max();
+        assert(x == 0xffffffffffffffff);
+        assert(api.u64_identity(x) == x);
+
+        x = api.i64_min();
+        assert(x == -9223372036854775808);
+        assert(api.i64_identity(x) == x);
+
+        x = api.i64_max();
+        assert(x == 9223372036854775807);
+        assert(api.i64_identity(x) == x);
+
+        x = api.i32_min();
+        assert(x == -2147483648);
+        assert(api.i32_identity(x) == x);
+
+        x = api.i32_max();
+        assert(x == 2147483647);
+        assert(api.i32_identity(x) == x);
+
+        x = api.u32_min();
+        assert(x == 0);
+        assert(api.u32_identity(x) == x);
+
+        x = api.u32_max();
+        assert(x == 0xffffffff);
+        assert(api.u32_identity(x) == x);
```

Comment:
this actually works? that's weird

Reply:
why wouldn't it?

Reply:
because it's passed as an i32 which would mean the wasm engine has no way of knowing it should interpret it as an unsigned number

Reply:
https://github.com/cloudpeers/ffi-gen/pull/33/files#diff-7f255e91d1075c4b17b67e15052b7562fb54c94bf7a7cbec359256343bf25ad6R388-R390
should fix it?

Reply:
that's only activated for u64/i64 right? I saw that, was curious about u32.

Reply:
I mean that
```rust
            Instr::LiftNum(r#in, out, NumType::U32) => {
                quote!(const #(self.var(out)) = #(self.var(r#in)) >>> 0;)
            }
```

Reply:
interesting, missed that. right shift by zero would be undefined behaviour in c and an error in rust. but seems to work.

Reply:
stole it from wasm-bindgen.